File: backend/app/domain/models/program_attempt.py
"""Minimal display proof for a confirmed first-party program attempt."""
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class ProgramAttemptView(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    version: Literal[1] = 1
    identity: str = Field(pattern=r"^[0-9a-f]{64}$")
    state: Literal["succeeded", "failed"]
    returncode: int

    @model_validator(mode="after")
    def consistent_status(self):
        if (self.returncode == 0) != (self.state == "succeeded"):
            raise ValueError("Program attempt status must agree with its confirmed exit code")
        return self

File: frontend/src/types/programAttempt.ts
/** Server-owned projection of a confirmed first-party execution receipt.
 * Identity is opaque: never reconstruct it from paths, labels or tool output.
 */
export interface ProgramAttemptView {
  version: 1;
  identity: string;
  state: 'succeeded' | 'failed';
  returncode: number;
}
